clients/spark/core/src/main/scala/io/treeverse/clients/LakeFSInputFormat.scala

```diff
@@ -1,74 +1,133 @@
+import scala.collection.mutable.ListBuffer
+import com.google.protobuf.ByteString
+import org.slf4j.{Logger, LoggerFactory}
 
-class GravelerSplit(var range: RangeData, var path: Path) extends InputSplit with Writable {
-  def this() = this(null, null)
+object GravelerSplit {
+  val logger: Logger = LoggerFactory.getLogger(getClass.toString)
+}
+class GravelerSplit(var range: RangeData, var path: Path, var rangeID: String, var byteSize: Long)
+    extends InputSplit
+    with Writable {
+  import GravelerSplit._
+  def this() = this(null, null, null, 0L)
 
   override def write(out: DataOutput): Unit = {
-    val encodedRange = range.toByteArray
-    out.writeInt(encodedRange.length)
-    out.write(encodedRange)
+    out.writeLong(byteSize)
+    out.writeInt(rangeID.length)
+    out.writeChars(rangeID)
+    if (range != null) {
+      val encodedRange = range.toByteArray
+      out.writeInt(encodedRange.length)
+      out.write(encodedRange)
+    } else {
+      out.writeInt(0)
+    }
     val p = path.toString
     out.writeInt(p.length)
     out.writeChars(p)
   }
 
   override def readFields(in: DataInput): Unit = {
+    byteSize = in.readLong()
+    val rangeIDLength = in.readInt()
+
+    val rangeIDBuilder = new StringBuilder
+    for (_ <- 1 to rangeIDLength) {
+      rangeIDBuilder += in.readChar()
+    }
+    rangeID = rangeIDBuilder.toString
     val encodedRangeLength = in.readInt()
-    val encodedRange = new Array[Byte](encodedRangeLength)
-    in.readFully(encodedRange)
-    range = RangeData.parseFrom(encodedRange)
+    if (encodedRangeLength == 0) {
+      range = null
+    } else {
+      val encodedRange = new Array[Byte](encodedRangeLength)
+      in.readFully(encodedRange)
+      range = RangeData.parseFrom(encodedRange)
+    }
     val pathLength = in.readInt()
     val p = new StringBuilder
     for (_ <- 1 to pathLength) {
       p += in.readChar()
     }
     path = new Path(p.result)
+    logger.debug(s"Read split $this")
   }
-
-  override def getLength: Long = range.estimatedSize
+  override def getLength: Long = byteSize
 
   override def getLocations: Array[String] = Array.empty[String]
 
   override def getLocationInfo: Array[SplitLocationInfo] =
     Array.empty[SplitLocationInfo]
+
+  override def toString: String = s"GravelerSplit(range=$range, path=$path, rangeID=$rangeID, byteSize=$byteSize)"
 }
 
 class WithIdentifier[Proto <: GeneratedMessage with scalapb.Message[Proto]](
     val id: Array[Byte],
-    val message: Proto
+    val message: Proto,
+    val rangeID: String
 ) {}
 
+object EntryRecordReader {
+  val logger: Logger = LoggerFactory.getLogger(getClass.toString)
+}
 class EntryRecordReader[Proto <: GeneratedMessage with scalapb.Message[Proto]](
     companion: GeneratedMessageCompanion[Proto]
 ) extends RecordReader[Array[Byte], WithIdentifier[Proto]] {
+  import EntryRecordReader._
   var it: SSTableIterator[Proto] = _
   var item: Item[Proto] = _
-
+  var gravelerSplit: GravelerSplit = _
```

Comment:
AFAICS only `rangeID` is actually needed outside of `initialize`.  Could we keep just that on the recordreader?